### PR TITLE
Restrict user overrides for remote store related index settings

### PR DESCRIPTION
**Pull request by linuxpi**
### Short Description
This change ensures user cannot tinker with any remote store related index settings. This will ensure our clusters remain homogeneous i.e. either the cluster has all remote enabled indices or no remote enabled indices.

### Detailed Description
With this change we are restricting the following index level settings:
- index.remote_store.enabled
- index.remote_store.segment.repository
- index.remote_store.translog.repository

user will not be able to override these settings while creating an index. 

If following cluster level settings are set then we will inject the above settings with the corresponding cluster level settings mentioned below.
- cluster.remote_store.enabled
- cluster.remote_store.segment.repository
- cluster.remote_store.translog.repository

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8770

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java [hunk 1]
```diff
@@ -90,18 +90,7 @@
...
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (andrross) [inline server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java:93]: No wildcard imports please

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -951,47 +949,39 @@ private static void updateReplicationStrategy(Settings.Builder settingsBuilder,
...
+        if (INDEX_REMOTE_STORE_ENABLED_SETTING.exists(requestSettings) == true) {
+            overriddenSettings.add(SETTING_REMOTE_STORE_ENABLED);
+        }
+        if (INDEX_REMOTE_SEGMENT_STORE_REPOSITORY_SETTING.exists(requestSettings) == true) {
+            overriddenSettings.add(SETTING_REMOTE_SEGMENT_STORE_REPOSITORY);
+        }
+        if (INDEX_REMOTE_TRANSLOG_REPOSITORY_SETTING.exists(requestSettings) == true) {
+            overriddenSettings.add(SETTING_REMOTE_TRANSLOG_STORE_REPOSITORY);
+        }
+        return overriddenSettings;
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:984]: Style nitpick, but I think you can write this as the following which in my opinion is bit less verbose and more clear:

```
return Stream.of(INDEX_REMOTE_STORE_ENABLED_SETTING,
        INDEX_REMOTE_TRANSLOG_REPOSITORY_SETTING,
        INDEX_REMOTE_SEGMENT_STORE_REPOSITORY_SETTING)
    .filter(setting -> setting.exists(requestSettings))
    .map(Setting::getKey)
    .collect(toList());
```

**Comment by andrross** [comment]
Is there an issue to link to this PR? Also, have we started writing the documentation for configuring the remote store? I know things are changing significantly from the original behavior and it will be easier to understand and review with a good description of the desired end state.

File: server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java (hunk 1 above)
Comment (linuxpi) [reply server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java:93]: need to remove IDE auto corrects. Thanks for pointing out

**Comment by linuxpi** [comment]
> Is there an issue to link to this PR? Also, have we started writing the documentation for configuring the remote store? I know things are changing significantly from the original behavior and it will be easier to understand and review with a good description of the desired end state.

Sure @andrross. we already had an issue created for this. just added it along with description.

File: server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java [hunk 3]
```diff
@@ -1191,40 +1192,11 @@ public void testvalidateIndexSettings() {
         threadPool.shutdown();
     }
 
-    public void testRemoteStoreNoUserOverrideConflictingReplicationTypeIndexSettings() {
```
Comment (linuxpi) [inline server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java:1194]: This UT will now be covered by ITs in CreateRemoteIndexIT.java

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+import static org.hamcrest.Matchers.greaterThan;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_REMOTE_SEGMENT_STORE_REPOSITORY;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_REMOTE_STORE_ENABLED;
+import static org.opensearch.indices.IndicesService.CLUSTER_REPLICATION_TYPE_SETTING;
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_STORE_ENABLED_SETTING;
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_SEGMENT_STORE_REPOSITORY_SETTING;
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_TRANSLOG_REPOSITORY_SETTING;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+
+public class RemoteRestoreSnapshotIT extends AbstractSnapshotIntegTestCase {
```
Comment (linuxpi) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:49]: These tests are copied from RestoreSnapshotIT to a new file as enabling remote store is only controlled at cluster level now. So single test file cannot create both remote and non remote indices

**Comment by linuxpi** [comment]
```
* What went wrong:
Execution failed for task ':distribution:bwc:staged:buildBwcLinuxTar'.
> Building 2.9.0 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/staged/build/bwc/checkout-2.9/distribution/archives/linux-tar/build/distributions/opensearch-min-2.9.0-SNAPSHOT-linux-x64.tar.gz
```

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 5]
```diff
@@ -951,47 +950,33 @@ private static void updateReplicationStrategy(Settings.Builder settingsBuilder,
...
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ROOT,
+                    "Cannot override [%s] settings when [%s] is set to [%s].",
+                    String.join("][", overriddenSettings),
+                    CLUSTER_REMOTE_STORE_ENABLED_SETTING.getKey(),
+                    isRemoteStoreClusterEnabled
+                )
+            );
+        }
```
[the remark this answers is not in the document]
Comment (linuxpi) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:965]: this is to restrict user override on remote store index level settings.
we dont want user to be able to set either of the following index level settings

```
index.remote_store.enabled
index.remote_store.segment.repository
index.remote_store.translog.repository
```

File: server/src/test/java/org/opensearch/repositories/blobstore/BlobStoreRepositoryRemoteIndexTests.java [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+import static org.hamcrest.Matchers.equalTo;
+import static org.opensearch.indices.IndicesService.CLUSTER_REPLICATION_TYPE_SETTING;
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_STORE_ENABLED_SETTING;
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_SEGMENT_STORE_REPOSITORY_SETTING;
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_TRANSLOG_REPOSITORY_SETTING;
+
+/**
+ * Tests for the {@link BlobStoreRepository} and its subclasses.
+ */
+@LuceneTestCase.SuppressFileSystems("ExtrasFS")
```
[the remark this answers is not in the document]
Comment (linuxpi) [reply server/src/test/java/org/opensearch/repositories/blobstore/BlobStoreRepositoryRemoteIndexTests.java:67]: yes, i'll remove

**Review by gbbafna — APPROVED** [review]
LGTM.

**Comment by linuxpi** [comment]
> @linuxpi : Can you please merge the backport PR #9025 to `2.x`. This is blocking other PRs which have changes on top of yours.

Working on fixing the one test that is failing. Looks like the code is diff on main vs 2.x . I've raise a PR for the fix #9084 will get it merged.

**Pull request merged**